### PR TITLE
added optional LATESTS flag to range,revrange,mrange,mrevrange,get co…

### PR DESCRIPTION
**Pull request by OfirMos**
…mmands

### CODE REVIEW

File: src/module.c [hunk 1]
```diff
@@ -1022,6 +1162,15 @@ int TSDB_mget(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
             RedisModule_ReplyWithArray(ctx, 0);
         }
         ReplyWithSeriesLastDatapoint(ctx, series);
+
+        // undo latest sample on LATEST param
+        if (should_finalize_last_bucket) {
+            if (srcSeries->totalSamples > 0) {
+                CompactionRule *rule = find_rule(srcSeries->rules, series->keyName);
+                SeriesDelRange(series, rule->startCurrentTimeBucket, rule->startCurrentTimeBucket);
```
Comment (danni-m) [inline src/module.c:1170]: @OfirMos why do we need to delete when LATEST is called?

File: src/module.c [hunk 2]
```diff
@@ -944,18 +1031,52 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 int TSDB_get(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     RedisModule_AutoMemory(ctx);
 
-    if (argc != 2) {
+    if (argc != 2 && argc != 3) {
```
Comment (danni-m) [inline src/module.c:1034]:
```suggestion
    if (argc < 2 || argc > 3) {
```

**Review by danni-m — CHANGES_REQUESTED** [review]
@OfirMos we shouldn't mutate the series when `LATEST` is requests, instead we should augment it by adding a proper parameter and function to the series interface.

File: src/module.c (hunk 1 above)
Comment (OfirMos) [reply src/module.c:1170]: @danni-m Yes, cause we closed the bucket while there might still arrive more samples that are belong to this bucket,
And as part of the latest implementation we wrote the last bucket to the downsampled series.

**Review by danni-m — APPROVED** [review]

File: src/libmr_integration.c [hunk 3]
```diff
@@ -669,7 +696,7 @@ Record *SeriesRecord_New(Series *series, timestamp_t startTimestamp, timestamp_t
             }
             break;
         }
-        if (series->funcs->GetLastTimestamp(chunk) > startTimestamp) {
+        if (series->funcs->GetLastTimestamp(chunk) >= startTimestamp) {
```
[the remark this answers is not in the document]
Comment (OfirMos) [reply src/libmr_integration.c:699]: Right, Not yet on master

File: src/libmr_integration.c [hunk 4]
```diff
@@ -678,6 +705,17 @@ Record *SeriesRecord_New(Series *series, timestamp_t startTimestamp, timestamp_t
             index++;
         }
     }
+
+    if (should_finalize_last_bucket(predicates, series)) {
+        Sample sample;
+        Sample *sample_ptr = &sample;
+        calculate_latest_sample(&sample_ptr, series);
+        if (sample_ptr && (sample.timestamp <= endTimestamp)) {
+            out->chunks[index] = out->funcs->NewChunk(128);
```
[the remark this answers is not in the document]
Comment (OfirMos) [reply src/libmr_integration.c:714]: Correct, this is to avoid the case of not enough space.

**Review by danni-m — APPROVED** [review]

**Pull request merged**